File: src/main/java/ml/duncte123/skybot/commands/fun/TagCommand.java
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017  Duncan "duncte123" Sterken & Ramid "ramidzkh" Khan & Maurice R S "Sanduhr32"
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.fun;

import ml.duncte123.skybot.Settings;
import ml.duncte123.skybot.Variables;
import ml.duncte123.skybot.objects.Tag;
import ml.duncte123.skybot.objects.command.Command;
import ml.duncte123.skybot.objects.command.CommandCategory;
import ml.duncte123.skybot.objects.command.CommandContext;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static me.duncte123.botcommons.messaging.MessageUtils.sendErrorWithMessage;
import static me.duncte123.botcommons.messaging.MessageUtils.sendMsg;
import static ml.duncte123.skybot.Settings.PATREON;
import static ml.duncte123.skybot.utils.CommandUtils.*;

public class TagCommand extends Command {

    private final Map<String, Tag> tagStore = new ConcurrentHashMap<>();

    public TagCommand(Variables variables) {
        this.category = CommandCategory.FUN;
        this.name = "tag";
        this.aliases = new String[]{
            "pasta",
            "tags",
            "t",
        };
        this.help = "Stores some text for later usage\n" +
            "Tags follow the same parsing as custom commands and the join/leave messages";
        this.usage = "<tag-name/raw/author/delete/create/help> [tag-name] [tag content]";

        variables.getDatabaseAdapter().loadTags((tags) -> {
            tags.forEach((tag) -> this.tagStore.put(tag.name, tag));

            return null;
        });
    }

    @Override
    public void execute(@Nonnull CommandContext ctx) {
        if ("tags".equalsIgnoreCase(ctx.getInvoke())) {
            sendTagsList(ctx);

            return;
        }

        final List<String> args = ctx.getArgs();

        if (args.isEmpty()) {
            sendMsg(ctx, "Missing arguments, check `" + ctx.getPrefix() + ctx.getInvoke() + " help`");

            return;
        }

        final String subCmd = args.get(0);

        if (args.size() == 1) {
            if ("help".equalsIgnoreCase(subCmd) || "?".equalsIgnoreCase(subCmd)) {
                sendTagHelp(ctx);

                return;
            }

            if ("list".equalsIgnoreCase(subCmd)) {
                sendTagsList(ctx);

                return;
            }
        }

        if (args.size() == 2 && handleTwoArgs(ctx, subCmd)) {
            return;
        }

        if ("create".equalsIgnoreCase(subCmd) || "new".equalsIgnoreCase(subCmd)) {
            createTag(ctx);

            return;
        }

        if (this.tagStore.containsKey(subCmd)) {
            sendTag(ctx, subCmd);

            return;
        }

        sendMsg(ctx, "Unknown tag `" + subCmd + "`, check `" + ctx.getPrefix() + ctx.getInvoke() + " help`");
    }

    private boolean handleTwoArgs(CommandContext ctx, String subCmd) {
        final String tagName = ctx.getArgs().get(1);

        if ("author".equalsIgnoreCase(subCmd) || "owner".equalsIgnoreCase(subCmd) || "who".equalsIgnoreCase(subCmd)) {
            sendTagOwner(ctx, tagName);

            return true;
        }

        if ("delete".equalsIgnoreCase(subCmd) || "remove".equalsIgnoreCase(subCmd)) {
            removeTag(ctx, tagName);

            return true;
        }

        if ("raw".equalsIgnoreCase(subCmd)) {
            sendTagRaw(ctx, tagName);

            return true;
        }

        return false;
    }

    private void sendTag(CommandContext ctx, String subCmd) {
        final String parsed = parseJagTag(ctx, this.tagStore.get(subCmd).content);

        if (parsed.length() > 2000) {
            sendErrorWithMessage(ctx.getMessage(), "Error: output is over 2000 character limit");

            return;
        }

        sendMsg(ctx, parsed);
    }

    private void sendTagHelp(CommandContext ctx) {
        final String invoke = ctx.getInvoke();
        final String prefix = ctx.getGuildSettings().getCustomPrefix();
        final String message = String.format(
            """
                Tag help:
                \t`%1$s%2$s help` => Shows this message
                \t`%1$s%2$s list` => Gives you a list of all the tags
                \t`%1$s%2$s author <tag name>` => Shows the owner of a tag
                \t`%1$s%2$s raw <tag name>` => Shows raw content of a tag
                \t`%1$s%2$s delete <tag name>` => Deletes a tag
                \t`%1$s%2$s create <tag name> <tag content>` => Creates a new tag""",
            prefix,
            invoke
        );

        sendMsg(ctx, message);
    }

    private void sendTagRaw(CommandContext ctx, String tagName) {
        if (!this.tagStore.containsKey(tagName)) {
            sendMsg(ctx, "That tag does not exist");

            return;
        }

        sendMsg(ctx, "```\n" + this.tagStore.get(tagName).content + "\n```");
    }

    private void sendTagsList(CommandContext ctx) {
        if (this.tagStore.isEmpty()) {
            sendMsg(ctx, "There are no tags in the system");

            return;
        }

        if (this.tagStore.size() < 100) {
            sendMsg(ctx, String.format("Here is the current tag list: `%s`", String.join("`, `", this.tagStore.keySet())));

            return;
        }

        final Permission perm = Permission.MESSAGE_ATTACH_FILES;

        if (!ctx.getSelfMember().hasPermission(ctx.getChannel(), perm)) {
            sendMsg(ctx, "I need the `" + perm.getName() + "` permission for this command to work");

            return;
        }

        final byte[] tagContent = ("[\"" + String.join("\", \"", this.tagStore.keySet()) + "\"]").getBytes();

        ctx.getChannel()
            .sendFile(tagContent, "tag_names.json")
            .append("Here is the current list of tags")
            .queue();
    }

    private void sendTagOwner(CommandContext ctx, String tagName) {
        if (!this.tagStore.containsKey(tagName)) {
            sendMsg(ctx, "That tag does not exist");

            return;
        }

        final long ownerId = this.tagStore.get(tagName).owner_id;
        final User user = ctx.getShardManager().getUserById(ownerId);
        final String userTag = user == null ? "UnknownUser#0000" : user.getAsTag();

        sendMsg(ctx, String.format("`%s` was created by `%s`", tagName, userTag));
    }

    private void removeTag(CommandContext ctx, String tagName) {
        if (!this.tagStore.containsKey(tagName)) {
            sendMsg(ctx, "That tag does not exist");

            return;
        }

        final Tag tag = this.tagStore.get(tagName);

        if (ctx.getAuthor().getIdLong() != tag.owner_id && !isDev(ctx.getAuthor())) {
            sendMsg(ctx, "You do not own that tag");

            return;
        }

        ctx.getDatabaseAdapter().deleteTag(tag, (success, reason) -> {
            if (!success) {
                sendMsg(ctx, "Failed to remove tag: " + reason);

                return null;
            }

            this.tagStore.remove(tag.name);

            sendMsg(ctx, String.format("Tag `%s` deleted", tag.name));

            return null;
        });
    }

    private void createTag(CommandContext ctx) {
        if (!isTagPatron(ctx.getMember())) {
            sendMsg(ctx, "Unfortunately only our tag tier patrons are able to create tags.\n" +
                "You can become one for $5/month over at <" + PATREON + '>');

            return;
        }

        final List<String> args = ctx.getArgs();
        final String tagName = args.get(1);

        if (this.tagStore.containsKey(tagName)) {
            sendMsg(ctx, "This tag already exists");

            return;
        }

        if (tagName.length() > 10) {
            sendMsg(ctx, "Max length for the name is 10 characters");

            return;
        }

        final Tag newTag = new Tag(
            tagName,
            ctx.getArgsRaw().split("\\s+", 3)[2],
            ctx.getAuthor().getIdLong()
        );

        ctx.getDatabaseAdapter().createTag(newTag, (success, reason) -> {
            if (!success) {
                sendMsg(ctx, "Failed to create tag: " + reason);

                return null;
            }

            this.tagStore.put(tagName, newTag);

            sendMsg(ctx, String.format("Tag `%s` created", tagName));

            return null;
        });
    }

    private boolean isTagPatron(Member member) {
        final User user = member.getUser();

        if (isUserTagPatron(user)) {
            return true;
        }

        if (member.getGuild().getIdLong() != Settings.SUPPORT_GUILD_ID) {
            return false;
        }

        final boolean hasRole = member.getRoles()
            .stream()
            .map(Role::getIdLong)
            .anyMatch((it) -> it == Settings.TAG_PATRONS_ROLE);

        if (hasRole) {
            TAG_PATRONS.add(user.getIdLong());
        }

        return hasRole;
    }
}

File: src/main/kotlin/ml/duncte123/skybot/commands/fun/TagCommand.kt
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017 - 2018  Duncan "duncte123" Sterken & Ramid "ramidzkh" Khan & Maurice R S "Sanduhr32"
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.`fun`

import ml.duncte123.skybot.objects.command.Command
import ml.duncte123.skybot.objects.command.CommandCategory
import ml.duncte123.skybot.utils.MessageUtils
import ml.duncte123.skybot.utils.Variables
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent

class TagCommand : Command() {

    init {
        this.category = CommandCategory.FUN
    }

    override fun executeCommand(invoke: String, args: Array<out String>, event: GuildMessageReceivedEvent) {
        MessageUtils.sendMsg(event, "The tag feature has been disabled, please use custom commands from this point forward")
        /*val helpMessage = MessageBuilder()
                .appendCodeBlock("Tag help: \n" +
                        "\t$PREFIX\u200B$invoke help: shows this \n" +
                        "\t$PREFIX\u200B$invoke list: lists all the tags \n" +
                        "\t$PREFIX\u200B$invoke delete: removes a tag \n" +
                        "\t$PREFIX\u200B$invoke author: displays who made the tag \n" +
                        "\t$PREFIX\u200B$invoke create: make a new tag \n", getLang()).build()
        if (args.isEmpty()) {
            MessageUtils.sendMsg(event, helpMessage)
        } else if (args.size == 1) {

            if (args[0] == "help" || args[0] == "?") {
                MessageUtils.sendMsg(event, helpMessage)
            } else if (args[0] == "list") {
                MessageUtils.sendMsg(event, "Here is a list of all the tags: `${StringUtils.join(TagUtils.tagsList.keys, "`, `")}`")
            } else {
                if (!TagUtils.tagsList.containsKey(args[0])) {
                    MessageUtils.sendMsg(event, "The tag `${args[0]}` does not exist.")
                    return
                }

                MessageUtils.sendMsg(event, """${TagUtils.tagsList[args[0]]!!.text}
                    |
                    |
                    |**WARNING!!!**
                    |This feature will be removed soon, to prevent your tags from being lost we suggest that you create custom commands for them
                    |You can find more info about using the custom commands by using `${PREFIX}customcommands help`
                """.trimMargin())
            }

        } else if (args.size == 2) {
            if (args[0] == "who" || args[0] == "author") {
                if (!TagUtils.tagsList.containsKey(args[1])) {
                    MessageUtils.sendMsg(event, "The tag `${args[1]}` does not exist.")
                    return
                }
                val t = TagUtils.tagsList[args[1]]

                MessageUtils.sendMsg(event, "The tag `${t!!.name}` is created by `${t.author}`.")
            } else if (args[0] == "delete" || args[0] == "remove") {
                if (!TagUtils.tagsList.containsKey(args[1])) {
                    MessageUtils.sendMsg(event, "The tag `${args[1]}` does not exist.")
                    return
                }

                val t = TagUtils.tagsList[args[1]]
                if (t!!.authorId != event.author.id) {
                    MessageUtils.sendMsg(event, "You do not own this tag.")
                    return
                }
                if (TagUtils.deleteTag(t)) {
                    MessageUtils.sendMsg(event, "Tag `${args[1]}` has been deleted successfully")
                } else {
                    MessageUtils.sendMsg(event, "Failed to delete this tag")
                }

            }
        } else if (args.size >= 3 && (args[0] == "create" || args[0] == "new")) {
            if (TagUtils.tagsList.containsKey(args[1])) {
                MessageUtils.sendMsg(event, "The tag `${args[1]}` already exist.")
                return
            } else if (args[1].length > 10) {
                MessageUtils.sendMsg(event, "Tag name is too long.")
                return
            } else if (args[1].contains("who") || args[1].contains("author") || args[1].contains("help") || args[1].contains("list") || args[1].contains("?") || args[1].contains("delete") || args[1].contains("remove")) {
                MessageUtils.sendMsg(event, "The tag name can't be `${args[1]}`")
                return
            }
            val newTagContent: List<String> = event.message.contentRaw.replaceFirst(Pattern.quote(PREFIX), "").split(" ")
            if (TagUtils.registerNewTag(event.author, Tag(
                            TagUtils.tagsList.keys.size + 1,
                            String.format("%#s", event.author),
                            event.author.id,
                            args[1],
                            StringUtils.join(Arrays.copyOfRange(newTagContent.toTypedArray(), 3, newTagContent.size), " ")))) {
                MessageUtils.sendMsg(event, "Tag added successfully.")

            } else {
                MessageUtils.sendMsg(event, "Failed to add tag.")
            }
            MessageUtils.sendMsg(event, "Creating new tags has been disabled due to the custom command system replacing it.\n" +
                    "Make sure to convert your tags to the new custom commands before {DATE SOON}")
        }*/
    }

    override fun help() = "Save it in a tag\n" +
            "Usage: `$PREFIX$name <tag_name/author/delete/create/help> [tag_name] [tag contents]`"

    override fun getName() = "tag"

    override fun getAliases() = arrayOf("pasta", "tags", "t")

    private fun getLang(): String {
        when (Variables.RAND.nextInt(4)) {
            0 -> return "YAML"
            1 -> return "ldif"
            2 -> return "PHP"
            3 -> return "CSS"
        }
        return "ldif"
    }
}